### PR TITLE
Fix #2772: Special case Devalify for java.lang.System.*

### LINKED ISSUE #2772: Local optimisations inline too eagerly: System.out is final but can still be mutated by System.setOut
```scala
import java.io.OutputStream

object Test {
  def main(args: Array[String]): Unit = {
    val oldOut = System.out
    System.setOut(null)
    val a = () => foo(oldOut)
    a()
  }

  def foo(out: OutputStream): Unit = {
    out.write(0)
  }
}
```
Without `-optimise` this code works, with `-optimise` we get a NullPointerException.
Without `-optimise`:
```scala
result of try/inout.scala after genBCode:

package <empty> {
  @scala.annotation.internal.SourceFile("try/inout.scala") final module class 
    Test$
   extends Object { 
    def <init>(): Unit = 
      {
        super()
        ()
      }
    def main(args: String[]): Unit = 
      {
        val oldOut: java.io.PrintStream = System.out
        System.setOut(null)
        val a: Function0 = 
          closure(oldOut | Test.$anonfun$1:scala.compat.java8.JFunction0$mcV$sp)
        {
          a.apply()
          ()
        }
      }
    def foo(out: java.io.OutputStream): Unit = out.write(0)
    private def $anonfun$1(oldOut$1: java.io.PrintStream): Unit = 
      Test.foo(oldOut$1)
  }
  final lazy module val Test: Test$ = new Test$()
}
```
With `-optimise`:
```scala
result of try/inout.scala after genBCode:

package <empty> {
  @scala.annotation.internal.SourceFile("try/inout.scala") final module class 
    Test$
   extends Object { 
    def <init>(): Unit = 
      {
        super()
        ()
      }
    def main(args: String[]): Unit = 
      {
        System.setOut(null)
        val a: Function0 = 
          closure(Test.$anonfun$1:scala.compat.java8.JFunction0$mcV$sp)
        {
          a.apply()
          ()
        }
      }
    def foo(out: java.io.OutputStream): Unit = out.write(0)
    private def $anonfun$1(): Unit = Test.foo(System.out)
  }
  final lazy module val Test: Test$ = new Test$()
}
```
Notice that the anonymous function now calls `System.out` instead of `out`.

This happened in the Dotty Language Server (https://github.com/lampepfl/dotty/blob/master/language-server/src/dotty/tools/languageserver/Main.scala#L32-L35), the result is that outside projects cannot use the IDE with `scalaVersion := 0.2.0-bin-20170614-8ddfcaf-NIGHTLY`.

### PR DESCRIPTION
**Pull request by OlivierBlanvillain**
@smarter thanks for the nice minimization!

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/localopt/Devalify.scala [hunk 1]
```diff
@@ -194,13 +198,22 @@ class Devalify extends Optimisation {
...
-      readingOnlyVals(qual)
+      if (t.symbol == defn.SystemModule) {
+        // System.in is static final fields that, for legacy reasons, must be
+        // allowed to be changed by the methods System.setIn...
+        // https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.5.4
+        false
+      } else
+        readingOnlyVals(qual)
+
     case t: Ident if !t.symbol.is(Mutable) && !t.symbol.is(Method) && !t.symbol.info.dealias.isInstanceOf[ExprType] =>
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/localopt/Devalify.scala]: Nothing says that there might not be an `Ident` corresponding to `System.in`, and in that case you would still inline it. Maybe add a `isReallyMutable` instead that you would use everywhere?

File: compiler/src/dotty/tools/dotc/transform/localopt/Devalify.scala [hunk 2]
```diff
@@ -182,8 +184,10 @@ class Devalify extends Optimisation {
       if (isGetterOfAImmutableField || isAccessingProductField || isImmutableCaseAccessor)
         readingOnlyVals(rec)
       else false
+
     case Select(rec, _) if t.symbol.is(Method) =>
-      if (t.symbol.isGetter && !t.symbol.is(Mutable)) readingOnlyVals(rec) // getter of a immutable field
+      if (t.symbol.isGetter && !t.symbol.is(Mutable))
+        readingOnlyVals(rec) // getter of a immutable field
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/localopt/Devalify.scala]: typo: a -> an

**Comment by smarter** [comment]
This also won't work if someone uses reflection to mutate a final field. I suggest creating a documentation page for `-optimize` to list all the things you're not allowed to do in your code for it to work.

File: compiler/src/dotty/tools/dotc/transform/localopt/Devalify.scala (hunk 1 above)
Comment (OlivierBlanvillain) [inline compiler/src/dotty/tools/dotc/transform/localopt/Devalify.scala]: The implementation of this case calls `desugarIdent` and recuse, I think the case where the `Ident` is `System.in` is covered.

@DarkDimius Could you explain why we even need this `desugarIdent`, why not use `Select` everywhere directly? Also where would be the proper was to move that function to factor out the version that's currently copy pasted in the backend?

**Comment by OlivierBlanvillain** [comment]
I went thought all usages of `Mutable` inside the `localopt` package and I found another potential but in `DropGoodCasts`, thanks for the suggestion! (All other uses of is(Mutable) shouldn't be affect by this System.stuff because it checks on safe things such as ValDefs).

**Comment by OlivierBlanvillain** [comment]
(I will later properly populate this documentation with content from the semester project report)

File: compiler/src/dotty/tools/dotc/transform/localopt/Devalify.scala [hunk 3]
```diff
@@ -164,7 +164,17 @@ class Devalify extends Optimisation {
     case _ => t
   }
 
-  def readingOnlyVals(t: Tree)(implicit ctx: Context): Boolean = dropCasts(t) match {
+  def readingOnlyVals(t: Tree)(implicit ctx: Context): Boolean = {
+    def isGetterOfAImmutableField = t.symbol.isGetter && !t.symbol.is(Mutable)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/localopt/Devalify.scala]: Of _An_ ImmutableField

File: compiler/src/dotty/tools/dotc/transform/localopt/Simplify.scala [hunk 4]
```diff
@@ -146,4 +146,15 @@ object Simplify {
       case _ => None
     }
   }
+
+  // System.in is static final fields that, for legacy reasons, must be
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/localopt/Simplify.scala]: is -> is a
fields -> field

This description is a bit confusing: it starts by talking just about System.in and then mention the rest.
The name of the method is also not great, it's too easy to confuse with `is(Mutable)`, I would call it something like `isEffectivelyMutable`, `isReallyMutable` or `isChanging`

File: compiler/src/dotty/tools/dotc/transform/localopt/Simplify.scala [hunk 5]
```diff
@@ -146,4 +146,15 @@ object Simplify {
...
+  // allowed to be changed by the methods System.setIn. `isMutable` is true
+  // is the field is Mutable or if it's a field of java.lang.System.
+  // https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.5.4
+  def isMutable(t: Tree)(implicit ctx: Context): Boolean = t match {
+    case _ if t.symbol.is(Mutable) => true
+    case s: Symbol => (s.symbol == defn.SystemModule)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/localopt/Simplify.scala]: I think you want s.owner.symbol here.

File: docs/docs/reference/optimisations.md [hunk 6]
```diff
@@ -0,0 +1,8 @@
+---
+layout: doc-page
+title: "Local Optimisations"
+---
+
+The Dotty compiler implements several local optimisations to generate faster bytecode. These optimisations can be enabled by compiling with the `-optimise` flag. The current best source of documentation about what tree transformations are performed under `-optimise` is the Scaladoc classes in the [localopt package](https://github.com/lampepfl/dotty/tree/master/compiler/src/dotty/tools/dotc/transform/localopt).
+
+Local optimisations assumes no use of Java Reflection to mutate static final fields.
```
Comment (smarter) [inline docs/docs/reference/optimisations.md]: What about non-static final fields, are those safe to mutate?

File: compiler/src/dotty/tools/dotc/transform/localopt/Simplify.scala [hunk 7]
```diff
@@ -146,4 +146,15 @@ object Simplify {
...
     }
   }
+
+  /** Is this tree mutable, or java.lang.System.{in, out, err}? These three
+   *  System members are the only static final fields that are mutable.
+   *  See https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.5.4
+   */
+  def isEffectivelyMutable(t: Tree)(implicit ctx: Context): Boolean = t match {
+    case _ if t.symbol.is(Mutable) => true
+    case s: Select => (s.symbol == defn.SystemModule)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/localopt/Simplify.scala]: This still looks wrong to me, `s` may be `System.in`, so you need to look at the owner

**Pull request merged**